File: client_examples_async.js
/**
 * Cliente JavaScript/TypeScript para API Assíncrona LangChain
 * ============================================================
 * 
 * Exemplos de como consumir a API assíncrona em aplicações frontend
 * (React, Angular, Vue, etc.) ou Node.js backend.
 */

// ============================================================================
// EXEMPLO 1: Cliente Básico (Vanilla JavaScript)
// ============================================================================

class LangChainAsyncClient {
    constructor(apiUrl, maxWaitTime = 300000) {
        this.apiUrl = apiUrl.replace(/\/$/, ''); // Remove trailing slash
        this.maxWaitTime = maxWaitTime; // 5 minutos em ms
    }

    /**
     * Envia consulta e retorna Promise com resultado
     */
    async query(clienteId, dadosCliente, pergunta, modo = 'segmento') {
        try {
            // 1. Enviar requisição
            const requestId = await this._sendRequest(
                clienteId, 
                dadosCliente, 
                pergunta, 
                modo
            );
            
            if (!requestId) {
                throw new Error('Falha ao enviar requisição');
            }

            // 2. Aguardar resultado com polling
            return await this._pollWithBackoff(requestId);

        } catch (error) {
            console.error('❌ Erro na consulta:', error);
            throw error;
        }
    }

    async _sendRequest(clienteId, dadosCliente, pergunta, modo) {
        const response = await fetch(`${this.apiUrl}/query`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({
                cliente_id: clienteId,
                dados_cliente: dadosCliente,
                pergunta: pergunta,
                modo: modo
            })
        });

        if (!response.ok) {
            throw new Error(`HTTP ${response.status}: ${response.statusText}`);
        }

        const data = await response.json();
        console.log(`✅ Requisição enviada: ${data.request_id}`);
        return data.request_id;
    }

    async _pollWithBackoff(requestId) {
        const startTime = Date.now();
        let waitTime = 1000; // Começar com 1 segundo
        const maxWait = 10000; // Máximo 10 segundos

        while (true) {
            // Verificar timeout
            const elapsed = Date.now() - startTime;
            if (elapsed > this.maxWaitTime) {
                throw new Error(`Timeout após ${elapsed / 1000}s`);
            }

            // Consultar status
            const response = await fetch(`${this.apiUrl}/status/${requestId}`);
            
            if (!response.ok) {
                throw new Error(`HTTP ${response.status}: ${response.statusText}`);
            }

            const data = await response.json();
            const status = data.status;

            if (status === 'COMPLETED') {
                console.log(`✅ Concluído em ${elapsed / 1000}s`);
                return data.result;
            }

            if (status === 'FAILED') {
                throw new Error(data.error || 'Processamento falhou');
            }

            // Aguardar antes da próxima consulta
            console.log(`⏳ Status: ${status} (aguardando ${waitTime / 1000}s)`);
            await this._sleep(waitTime);

            // Exponential backoff
            waitTime = Math.min(waitTime * 2, maxWait);
        }
    }

    _sleep(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }
}


// ============================================================================
// EXEMPLO 2: React Component com UI
// ============================================================================

/**
 * Hook React para usar a API assíncrona
 */
function useLangChainQuery() {
    const [loading, setLoading] = React.useState(false);
    const [progress, setProgress] = React.useState(null);
    const [result, setResult] = React.useState(null);
    const [error, setError] = React.useState(null);

    const query = async (clienteId, dadosCliente, pergunta, modo = 'segmento') => {
        setLoading(true);
        setProgress('Enviando requisição...');
        setResult(null);
        setError(null);

        const apiUrl = process.env.REACT_APP_API_URL;
        const client = new LangChainAsyncClient(apiUrl);

        try {
            // Enviar requisição
            const response = await fetch(`${apiUrl}/query`, {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({
                    cliente_id: clienteId,
                    dados_cliente: dadosCliente,
                    pergunta: pergunta,
                    modo: modo
                })
            });

            const { request_id } = await response.json();
            setProgress(`Processando... (ID: ${request_id})`);

            // Polling com atualizações de progresso
            let waitTime = 1000;
            const maxWait = 10000;
            const startTime = Date.now();

            while (true) {
                const statusResponse = await fetch(`${apiUrl}/status/${request_id}`);
                const statusData = await statusResponse.json();

                const elapsed = Math.floor((Date.now() - startTime) / 1000);
                setProgress(`${statusData.status}... (${elapsed}s)`);

                if (statusData.status === 'COMPLETED') {
                    setResult(statusData.result);
                    setLoading(false);
                    break;
                }

                if (statusData.status === 'FAILED') {
                    throw new Error(statusData.error || 'Processamento falhou');
                }

                await new Promise(resolve => setTimeout(resolve, waitTime));
                waitTime = Math.min(waitTime * 2, maxWait);
            }

        } catch (err) {
            setError(err.message);
            setLoading(false);
        }
    };

    return { loading, progress, result, error, query };
}

/**
 * Componente React de exemplo
 */
function ChatComponent() {
    const { loading, progress, result, error, query } = useLangChainQuery();
    const [pergunta, setPergunta] = React.useState('');

    const handleSubmit = async (e) => {
        e.preventDefault();
        
        await query(
            'C12345',
            {
                idade: 35,
                renda_mensal: 5000,
                saldo_medio: 8000,
                transacoes_mes: 15,
                score_credito: 680,
                num_produtos: 3
            },
            pergunta,
            'segmento'
        );
    };

    return (
        <div className="chat-container">
            <form onSubmit={handleSubmit}>
                <textarea
                    value={pergunta}
                    onChange={(e) => setPergunta(e.target.value)}
                    placeholder="Digite sua pergunta..."
                    disabled={loading}
                />
                <button type="submit" disabled={loading}>
                    {loading ? 'Processando...' : 'Enviar'}
                </button>
            </form>

            {loading && (
                <div className="loading">
                    <div className="spinner"></div>
                    <p>{progress}</p>
                </div>
            )}

            {error && (
                <div className="error">
                    ❌ Erro: {error}
                </div>
            )}

            {result && (
                <div className="result">
                    <h3>Resultado:</h3>
                    <p><strong>Segmento:</strong> {result.segmento}</p>
                    <p><strong>Resposta:</strong> {result.resposta}</p>
                </div>
            )}
        </div>
    );
}


// ============================================================================
// EXEMPLO 3: Angular Service
// ============================================================================

/**
 * Service Angular para consumir API assíncrona
 */
/*
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, interval, throwError } from 'rxjs';
import { switchMap, takeWhile, catchError, map } from 'rxjs/operators';

@Injectable({
  providedIn: 'root'
})
export class LangChainService {
  private apiUrl = environment.apiUrl;

  constructor(private http: HttpClient) {}

  query(
    clienteId: string,
    dadosCliente: any,
    pergunta: string,
    modo: string = 'segmento'
  ): Observable<any> {
    // 1. Enviar requisição
    return this.http.post<{ request_id: string }>(`${this.apiUrl}/query`, {
      cliente_id: clienteId,
      dados_cliente: dadosCliente,
      pergunta: pergunta,
      modo: modo
    }).pipe(
      // 2. Iniciar polling
      switchMap(({ request_id }) => this.pollStatus(request_id))
    );
  }

  private pollStatus(requestId: string): Observable<any> {
    return interval(2000).pipe( // Consultar a cada 2 segundos
      switchMap(() => this.http.get<any>(`${this.apiUrl}/status/${requestId}`)),
      takeWhile(response => {
        // Continuar enquanto estiver PENDING ou PROCESSING
        return response.status !== 'COMPLETED' && response.status !== 'FAILED';
      }, true), // Incluir o último valor
      map(response => {
        if (response.status === 'COMPLETED') {
          return response.result;
        }
        if (response.status === 'FAILED') {
          throw new Error(response.error || 'Processamento falhou');
        }
        return response; // Retornar status intermediário
      }),
      catchError(error => {
        console.error('Erro no polling:', error);
        return throwError(() => error);
      })
    );
  }
}
*/


// ============================================================================
// EXEMPLO 4: Node.js Backend (Express)
// ============================================================================

/**
 * Rota Express que usa a API assíncrona como proxy
 */
/*
const express = require('express');
const axios = require('axios');

const router = express.Router();
const LANGCHAIN_API_URL = process.env.LANGCHAIN_API_URL;

router.post('/consulta', async (req, res) => {
    const { cliente_id, dados_cliente, pergunta, modo } = req.body;

    try {
        // 1. Enviar para API assíncrona
        const { data: { request_id } } = await axios.post(
            `${LANGCHAIN_API_URL}/query`,
            { cliente_id, dados_cliente, pergunta, modo }
        );

        // 2. Retornar request_id imediatamente
        res.status(202).json({
            request_id,
            status: 'PENDING',
            message: `Consulte GET /consulta/${request_id} para ver o resultado`
        });

    } catch (error) {
        res.status(500).json({ error: error.message });
    }
});

router.get('/consulta/:request_id', async (req, res) => {
    const { request_id } = req.params;

    try {
        const { data } = await axios.get(
            `${LANGCHAIN_API_URL}/status/${request_id}`
        );

        res.json(data);

    } catch (error) {
        res.status(500).json({ error: error.message });
    }
});

module.exports = router;
*/


// ============================================================================
// EXEMPLO 5: TypeScript com Tipos
// ============================================================================

interface DadosCliente {
    idade: number;
    renda_mensal: number;
    saldo_medio: number;
    transacoes_mes: number;
    score_credito: number;
    num_produtos: number;
}

interface QueryRequest {
    cliente_id: string;
    dados_cliente: DadosCliente;
    pergunta: string;
    modo: 'segmento' | 'persona' | 'rag';
}

interface QueryResponse {
    request_id: string;
    status: 'PENDING';
    message: string;
}

interface StatusResponse {
    request_id: string;
    status: 'PENDING' | 'PROCESSING' | 'COMPLETED' | 'FAILED';
    result?: any;
    error?: string;
    created_at?: string;
    started_at?: string;
    completed_at?: string;
}

class TypedLangChainClient {
    constructor(
        private apiUrl: string,
        private maxWaitTime: number = 300000
    ) {
        this.apiUrl = apiUrl.replace(/\/$/, '');
    }

    async query(request: QueryRequest): Promise<any> {
        // 1. Enviar requisição
        const response = await fetch(`${this.apiUrl}/query`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(request)
        });

        if (!response.ok) {
            throw new Error(`HTTP ${response.status}`);
        }

        const { request_id }: QueryResponse = await response.json();

        // 2. Polling
        return this.pollStatus(request_id);
    }

    private async pollStatus(requestId: string): Promise<any> {
        const startTime = Date.now();
        let waitTime = 1000;
        const maxWait = 10000;

        while (true) {
            if (Date.now() - startTime > this.maxWaitTime) {
                throw new Error('Timeout');
            }

            const response = await fetch(`${this.apiUrl}/status/${requestId}`);
            const data: StatusResponse = await response.json();

            if (data.status === 'COMPLETED') {
                return data.result;
            }

            if (data.status === 'FAILED') {
                throw new Error(data.error || 'Failed');
            }

            await this.sleep(waitTime);
            waitTime = Math.min(waitTime * 2, maxWait);
        }
    }

    private sleep(ms: number): Promise<void> {
        return new Promise(resolve => setTimeout(resolve, ms));
    }
}


// ============================================================================
// EXEMPLO DE USO
// ============================================================================

// Vanilla JS
const client = new LangChainAsyncClient('https://seu-endpoint.execute-api.us-east-1.amazonaws.com');

client.query(
    'C12345',
    {
        idade: 35,
        renda_mensal: 5000,
        saldo_medio: 8000,
        transacoes_mes: 15,
        score_credito: 680,
        num_produtos: 3
    },
    'Quais produtos você recomenda?',
    'segmento'
)
.then(result => {
    console.log('✅ Resultado:', result);
})
.catch(error => {
    console.error('❌ Erro:', error);
});

// TypeScript
const typedClient = new TypedLangChainClient('https://seu-endpoint...');

(async () => {
    try {
        const result = await typedClient.query({
            cliente_id: 'C12345',
            dados_cliente: {
                idade: 35,
                renda_mensal: 5000,
                saldo_medio: 8000,
                transacoes_mes: 15,
                score_credito: 680,
                num_produtos: 3
            },
            pergunta: 'Quais produtos você recomenda?',
            modo: 'segmento'
        });

        console.log('✅ Resultado:', result);
    } catch (error) {
        console.error('❌ Erro:', error);
    }
})();

// Export para uso em módulos
if (typeof module !== 'undefined' && module.exports) {
    module.exports = {
        LangChainAsyncClient,
        TypedLangChainClient
    };
}

File: client_examples_async.py
"""
Exemplos de Cliente para API Assíncrona LangChain
===================================================

Este arquivo contém exemplos práticos de como consumir a nova API assíncrona
que resolve o problema de timeout de 30 segundos.
"""

import requests
import time
from typing import Dict, Optional
import json


# ============================================================================
# EXEMPLO 1: Cliente Básico com Polling Simples
# ============================================================================

def exemplo_basico():
    """Exemplo mais simples de uso da API assíncrona"""
    
    # URL da API (obter do Terraform output)
    API_URL = "https://seu-endpoint.execute-api.us-east-1.amazonaws.com"
    
    # 1. Enviar requisição assíncrona
    response = requests.post(
        f"{API_URL}/query",
        json={
            "cliente_id": "C12345",
            "dados_cliente": {
                "idade": 35,
                "renda_mensal": 5000,
                "saldo_medio": 8000,
                "transacoes_mes": 15,
                "score_credito": 680,
                "num_produtos": 3
            },
            "pergunta": "Quais produtos você recomenda?",
            "modo": "segmento"
        }
    )
    
    data = response.json()
    request_id = data["request_id"]
    print(f"✅ Requisição enviada: {request_id}")
    
    # 2. Fazer polling até completar
    while True:
        status_response = requests.get(f"{API_URL}/status/{request_id}")
        status_data = status_response.json()
        
        status = status_data["status"]
        print(f"⏳ Status: {status}")
        
        if status == "COMPLETED":
            print(f"✅ Resultado: {status_data['result']}")
            break
        elif status == "FAILED":
            print(f"❌ Erro: {status_data.get('error')}")
            break
        
        # Aguardar antes de consultar novamente
        time.sleep(5)


# ============================================================================
# EXEMPLO 2: Cliente com Exponential Backoff
# ============================================================================

class AsyncLangChainClient:
    """Cliente robusto com retry e exponential backoff"""
    
    def __init__(self, api_url: str, max_wait_time: int = 300):
        """
        Args:
            api_url: URL base da API
            max_wait_time: Tempo máximo de espera em segundos (padrão: 5 min)
        """
        self.api_url = api_url.rstrip("/")
        self.max_wait_time = max_wait_time
    
    def query(self, 
              cliente_id: str,
              dados_cliente: Dict,
              pergunta: str,
              modo: str = "segmento") -> Optional[Dict]:
        """
        Envia consulta e aguarda resultado com exponential backoff
        
        Returns:
            Resultado da consulta ou None se falhar
        """
        # 1. Enviar requisição
        request_id = self._send_request(cliente_id, dados_cliente, pergunta, modo)
        if not request_id:
            return None
        
        # 2. Aguardar com exponential backoff
        return self._poll_with_backoff(request_id)
    
    def _send_request(self, 
                      cliente_id: str,
                      dados_cliente: Dict,
                      pergunta: str,
                      modo: str) -> Optional[str]:
        """Envia requisição e retorna request_id"""
        try:
            response = requests.post(
                f"{self.api_url}/query",
                json={
                    "cliente_id": cliente_id,
                    "dados_cliente": dados_cliente,
                    "pergunta": pergunta,
                    "modo": modo
                },
                timeout=10
            )
            response.raise_for_status()
            
            data = response.json()
            request_id = data["request_id"]
            print(f"✅ Requisição enviada: {request_id}")
            return request_id
            
        except requests.exceptions.RequestException as e:
            print(f"❌ Erro ao enviar requisição: {e}")
            return None
    
    def _poll_with_backoff(self, request_id: str) -> Optional[Dict]:
        """
        Faz polling com exponential backoff
        
        Intervalo de consultas:
        - 1s, 2s, 4s, 8s, 10s, 10s, 10s...
        """
        start_time = time.time()
        wait_time = 1  # Começar com 1 segundo
        max_wait = 10  # Máximo de 10 segundos entre consultas
        
        while True:
            # Verificar timeout
            elapsed = time.time() - start_time
            if elapsed > self.max_wait_time:
                print(f"⏱️ Timeout após {elapsed:.1f}s")
                return None
            
            # Consultar status
            try:
                response = requests.get(
                    f"{self.api_url}/status/{request_id}",
                    timeout=5
                )
                response.raise_for_status()
                data = response.json()
                
                status = data["status"]
                
                if status == "COMPLETED":
                    print(f"✅ Concluído em {elapsed:.1f}s")
                    return data["result"]
                
                elif status == "FAILED":
                    error = data.get("error", "Erro desconhecido")
                    print(f"❌ Falha: {error}")
                    return None
                
                else:
                    print(f"⏳ Status: {status} (aguardando {wait_time}s)")
                    time.sleep(wait_time)
                    
                    # Exponential backoff
                    wait_time = min(wait_time * 2, max_wait)
                    
            except requests.exceptions.RequestException as e:
                print(f"⚠️ Erro ao consultar status: {e}")
                time.sleep(wait_time)


# ============================================================================
# EXEMPLO 3: Cliente Assíncrono (asyncio)
# ============================================================================

import asyncio
import aiohttp

class AsyncioLangChainClient:
    """Cliente totalmente assíncrono usando asyncio"""
    
    def __init__(self, api_url: str, max_wait_time: int = 300):
        self.api_url = api_url.rstrip("/")
        self.max_wait_time = max_wait_time
    
    async def query(self,
                    cliente_id: str,
                    dados_cliente: Dict,
                    pergunta: str,
                    modo: str = "segmento") -> Optional[Dict]:
        """Versão async do método query"""
        
        async with aiohttp.ClientSession() as session:
            # 1. Enviar requisição
            request_id = await self._send_request(
                session, cliente_id, dados_cliente, pergunta, modo
            )
            if not request_id:
                return None
            
            # 2. Aguardar com backoff
            return await self._poll_with_backoff(session, request_id)
    
    async def _send_request(self,
                           session: aiohttp.ClientSession,
                           cliente_id: str,
                           dados_cliente: Dict,
                           pergunta: str,
                           modo: str) -> Optional[str]:
        """Envia requisição (async)"""
        try:
            async with session.post(
                f"{self.api_url}/query",
                json={
                    "cliente_id": cliente_id,
                    "dados_cliente": dados_cliente,
                    "pergunta": pergunta,
                    "modo": modo
                },
                timeout=aiohttp.ClientTimeout(total=10)
            ) as response:
                response.raise_for_status()
                data = await response.json()
                request_id = data["request_id"]
                print(f"✅ Requisição enviada: {request_id}")
                return request_id
                
        except Exception as e:
            print(f"❌ Erro ao enviar requisição: {e}")
            return None
    
    async def _poll_with_backoff(self,
                                 session: aiohttp.ClientSession,
                                 request_id: str) -> Optional[Dict]:
        """Polling com backoff (async)"""
        start_time = time.time()
        wait_time = 1
        max_wait = 10
        
        while True:
            elapsed = time.time() - start_time
            if elapsed > self.max_wait_time:
                print(f"⏱️ Timeout após {elapsed:.1f}s")
                return None
            
            try:
                async with session.get(
                    f"{self.api_url}/status/{request_id}",
                    timeout=aiohttp.ClientTimeout(total=5)
                ) as response:
                    response.raise_for_status()
                    data = await response.json()
                    
                    status = data["status"]
                    
                    if status == "COMPLETED":
                        print(f"✅ Concluído em {elapsed:.1f}s")
                        return data["result"]
                    
                    elif status == "FAILED":
                        error = data.get("error", "Erro desconhecido")
                        print(f"❌ Falha: {error}")
                        return None
                    
                    else:
                        print(f"⏳ Status: {status}")
                        await asyncio.sleep(wait_time)
                        wait_time = min(wait_time * 2, max_wait)
                        
            except Exception as e:
                print(f"⚠️ Erro ao consultar status: {e}")
                await asyncio.sleep(wait_time)


# ============================================================================
# EXEMPLO 4: Múltiplas Requisições em Paralelo
# ============================================================================

async def processar_em_lote(clientes: list):
    """
    Processa múltiplos clientes em paralelo
    
    Args:
        clientes: Lista de dicts com dados dos clientes
    """
    API_URL = "https://seu-endpoint.execute-api.us-east-1.amazonaws.com"
    client = AsyncioLangChainClient(API_URL)
    
    # Criar tarefas para todos os clientes
    tasks = []
    for cliente in clientes:
        task = client.query(
            cliente_id=cliente["id"],
            dados_cliente=cliente["dados"],
            pergunta=cliente["pergunta"],
            modo=cliente.get("modo", "segmento")
        )
        tasks.append(task)
    
    # Executar todas em paralelo
    results = await asyncio.gather(*tasks)
    
    # Processar resultados
    for cliente, result in zip(clientes, results):
        if result:
            print(f"Cliente {cliente['id']}: ✅ {result['segmento']}")
        else:
            print(f"Cliente {cliente['id']}: ❌ Falhou")
    
    return results


# ============================================================================
# EXEMPLO DE USO
# ============================================================================

if __name__ == "__main__":
    # Configurar API URL (obter do Terraform)
    API_URL = "https://seu-endpoint.execute-api.us-east-1.amazonaws.com"
    
    # Exemplo 1: Cliente básico
    print("\n=== EXEMPLO 1: Cliente Básico ===")
    # exemplo_basico()
    
    # Exemplo 2: Cliente com retry
    print("\n=== EXEMPLO 2: Cliente com Exponential Backoff ===")
    client = AsyncLangChainClient(API_URL, max_wait_time=300)
    resultado = client.query(
        cliente_id="C12345",
        dados_cliente={
            "idade": 35,
            "renda_mensal": 5000,
            "saldo_medio": 8000,
            "transacoes_mes": 15,
            "score_credito": 680,
            "num_produtos": 3
        },
        pergunta="Quais produtos você recomenda?",
        modo="segmento"
    )
    
    if resultado:
        print(f"\n📊 Resultado Final:")
        print(json.dumps(resultado, indent=2, ensure_ascii=False))
    
    # Exemplo 3: Cliente asyncio
    print("\n=== EXEMPLO 3: Cliente Async ===")
    async_client = AsyncioLangChainClient(API_URL)
    # resultado_async = asyncio.run(async_client.query(...))
    
    # Exemplo 4: Processamento em lote
    print("\n=== EXEMPLO 4: Processamento em Lote ===")
    clientes = [
        {
            "id": "C001",
            "dados": {"idade": 25, "renda_mensal": 3000, "saldo_medio": 2000},
            "pergunta": "Quais produtos para mim?",
            "modo": "segmento"
        },
        {
            "id": "C002",
            "dados": {"idade": 45, "renda_mensal": 10000, "saldo_medio": 50000},
            "pergunta": "Investimentos recomendados?",
            "modo": "persona"
        },
        # ... mais clientes
    ]
    # resultados_lote = asyncio.run(processar_em_lote(clientes))
